src/server.js

```diff
@@ -3,50 +3,70 @@ import { StaticRouter } from 'react-router-dom';
 import express from 'express';
 import { renderToString } from 'react-dom/server';
 import Root from './client/components/Root';
+import { getCollection } from './api/collection';
+import { CollectionProvider } from './client/components/CollectionProvider';
 
 // eslint-disable-next-line import/no-dynamic-require
 const assets = require(process.env.RAZZLE_ASSETS_MANIFEST);
 
-const server = express();
-server
-  .disable('x-powered-by')
-  .use(express.static(process.env.RAZZLE_PUBLIC_DIR))
-  .get('/*', (req, res) => {
-    const context = {};
-    const markup = renderToString(
+const handleRender = (req, res, data) => {
+  const context = {};
+  const markup = renderToString(
+    <CollectionProvider data={data}>
       <StaticRouter context={context} location={req.url}>
         <Root />
-      </StaticRouter>,
-    );
+      </StaticRouter>
+    </CollectionProvider>,
+  );
 
-    if (context.url) {
-      res.redirect(context.url);
-    } else {
-      res.status(200).send(
-        `<!doctype html>
-    <html lang="">
-    <head>
-        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
-        <meta charset="utf-8" />
-        <title>Graasp</title>
-        <meta name="viewport" content="width=device-width, initial-scale=1">
-        ${
-          assets.client.css
-            ? `<link rel="stylesheet" href="${assets.client.css}">`
-            : ''
-        }
-        ${
-          process.env.NODE_ENV === 'production'
-            ? `<script src="${assets.client.js}" defer></script>`
-            : `<script src="${assets.client.js}" defer crossorigin></script>`
-        }
-    </head>
-    <body>
-        <div id="root">${markup}</div>
-    </body>
+  if (context.url) {
+    res.redirect(context.url);
+  } else {
+    res.status(200).send(
+      `<!doctype html>
+  <html lang="">
+  <head>
+      <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+      <meta charset="utf-8" />
+      <title>Graasp</title>
+      <meta name="viewport" content="width=device-width, initial-scale=1">
+      ${
+        assets.client.css
+          ? `<link rel="stylesheet" href="${assets.client.css}">`
+          : ''
+      }
+      ${
+        process.env.NODE_ENV === 'production'
+          ? `<script src="${assets.client.js}" defer></script>`
+          : `<script src="${assets.client.js}" defer crossorigin></script>`
+      }
+  </head>
+  <body>
+      <div id="root">${markup}</div>
+      <script>
+        // WARNING: See the following for security issues around embedding JSON in HTML:
+        // https://redux.js.org/recipes/server-rendering/#security-considerations
+        window.PRELOADED_STATE = ${JSON.stringify(data)?.replace(
```

Comment:
todo: ensure that this can only have data we allow

Reply:
You can leave this for a future PR.

Reply:
address in issue #15